Add EIP-5604: NFT Lien (#5604)
ethereum/EIPs: EIPS/eip-5604.md

---
eip: 5604

> title: NFT Lien

Reviewer: ```suggestion
title: NFT Liens
```

Author: Thank you for the suggetion.

We prefer to use the singular form to refer to a type of thing. Such as 
- Lien (instead of Liens) as in https://en.wikipedia.org/wiki/Lien
- Token (instead of Tokens) as in eip-20.md

Reviewer: Fair enough.

> description: Extend NFT to support lien

Reviewer: ```suggestion
description: Extend EIP-721 to support liens
```

author: Zainan Victor Zhou (@xinbenlv)
discussions-to: https://ethereum-magicians.org/t/creating-a-new-erc-proposal-for-nft-lien/10683
status: Draft
type: Standards Track
category: ERC
created: 2022-09-05
requires: 721
---

## Abstract

> This EIP introduce NFT lien, a form of security interest over an item of property to secure the recovery of liability or performance of some other obligation. It introduces interface to place and remove a lien, plus an event.

Reviewer: ```suggestion
This EIP introduces NFT liens, a form of security interest over an item of property to secure the recovery of liability or performance of some other obligation. It introduces an interface to place and removes a lien, plus an event.
```

## Motivation

> Lien is widely used for finance use cases, such as car and property liens. An use case for NFT is deed.

Reviewer: ```suggestion
Liens are widely used for finance use cases, such as car and property liens. An example use case for an NFT lien is for a deed.
```

> The EIP provides an interface to implement interface that perform the lien holding relationships.

Reviewer: ```suggestion
This EIP provides an interface to implement an interface that performs the lien holding relationships.
```

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> 1. Any compliant contract MUST implement [EIP-721](./eip-721.md), [EIP-165](./eip-165.md).

Reviewer: ```suggestion
1. Any compliant contract MUST implement [EIP-721](./eip-721.md), and [EIP-165](./eip-165.md).
```

2. Any compliant contract MUST implement the following interface:

```solidity
interface IERC_LIEN is EIP721, EIP165 {
    event OnLienPlaced(uint256 tokenId, address holder, bytes calldata extraParams)
    function addLienHolder(uint256 tokenId, address holder, bytes calldata extraParams);
    function removeLienHolder(uint256 tokenId, address holder, bytes calldata extraParams);
}
```

> 3. `addLienHolder` MUST throw error if

Reviewer: Missing end of sentence

Author: Thanks for catching! Will follow up. this was due to keystroke mistake

## Rationale

> 1. We only support NFT of [EIP-721](./eip-721.md) for simplicity and gas efficiency. We have not considered other ERCs which can left for future extension. For example, EIP-20 and EIP-1155 was not considered.

Reviewer: ```suggestion
1. We only support [EIP-721](./eip-721.md) NFTs for simplicity and gas efficiency. We have not considered other EIPs, which can be left for future extensions. For example, EIP-20 and EIP-1155 were not considered.
```

2. We choose separate "addLienHolder" and "removeLienHolder" instead of use a single `changeLienholder` with amount because we believe
the add or remove action are significantly different and usually require different Access Control,
for example, the token holder shall be able to add someone else as a lien holder but the lien holder of that token

3. We have not specified the "amount of debt" in this interface. We believe this is complex enough and worthy of an individual EIP by itself.

4. We have not specified how endorsement can be applied to allow holder to signal their approval for transfer or swapping. We believe this is complex enough and worthy of an individual EIP by itself.

## Backwards Compatibility

> The EIP is designed as an extension of EIP-721 and therefore compliant contract need to be fully complying with EIP-721.

Reviewer: ```suggestion
The EIP is designed as an extension of EIP-721 and therefore compliant contracts need to fully comply with EIP-721.
```
